### DISCUSSION TITLE
taro 怎么适用mqtt

### DISCUSSION THREAD

**Discussion by chenglei32870** [NervJS/taro#11164] [Q & A]
急 用mqtt 就报错

**Comment by Chen-jj** [comment]
报什么错，什么信息都没有，很难帮你

**Reply by chenglei32870** [reply]
```
getMonitorTokenByLessonId().then((res) => {
    const tokenObj = res.payload
    console.log(tokenObj)
    let username = 'Token|' + tokenObj.accessKey + '|' + tokenObj.instanceId
    let password = tokenObj.action + '|' + tokenObj.token

    let options = {
      username: username,
      password: password,
      clientId: tokenObj.clientId,
      keepalive: 90,
      connectTimeout: 3000,
    }
    console.log('mqtt connect options:' + JSON.stringify(options))
    const client = mqtt.connect(
      tokenObj.protocol + tokenObj.host + ':' + tokenObj.port,
      options,
    )
    client.on('connect', function () {
      console.log('connect success')
    })
    client.on('message', function (topic, message) {
      console.log('topic:' + topic + ' msg:' + message.toString())
    })
  `})```
```
这个是我的代码

**Reply by chenglei32870** [reply]
client.on('connect', function () {
      console.log('connect success')
    })
在connect 的时候报错了 报的错就是说websocket 没有构造函数

**Reply by chenglei32870** [reply]
`Unhandled promise rejection TypeError: WebSocket is not a constructor
    at index.js:770
    at Object.t.exports (index.js:767)
    at T.streamBuilder (index.js:967)
    at T._setupStream (index.js:387)
    at new T (index.js:373)
    at Function.u [as connect] (index.js:966)
    at index.js:5746
    at VM1495 WAService.js:2
    at n (VM1495 WAService.js:2)(env: Windows,mp,1.05.2106250; lib: 2.21.2)`

**Reply by Chen-jj** [reply]
> client.on('connect', function () { console.log('connect success') }) 在connect 的时候报错了 报的错就是说websocket 没有构造函数

@chenglei32870 那是当然，小程序的 WebSocket API 和 Web 不一样。建议先看看小程序文档。

如果你要用这个库，可以参考 Taro React DevTools 插件是怎么封装 Web 版本的 WebSocket API 的：https://github.com/NervJS/taro/blob/next/packages/taro-plugin-react-devtools/src/websocket.ts
